### PR TITLE
[tool] cleanup bundle builder a bit

### LINKED ISSUE #80732: [tool] cleanup bundle building a bit
The [`BundleBuilder`](https://github.com/flutter/flutter/blob/ce18d702e90d3dff9fee53d61a770c94f14f2811/packages/flutter_tools/lib/src/bundle.dart#L80) currently accepts some parameters that are already contained in `buildInfo`. For example: `trackWidgetCreation`, `extraFrontEndOptions`, `extraGenSnapshotOptions`, `fileSystemRoots`, `fileSystemScheme` and `treeShakeIcons`. Some of these are also not applied correctly to the build.

Also the `buildWithAssemble` shim is solely used for BundleBuilder, so we may as well put that functionality into BundleBuilder. (Maybe helps keeping the behaviour consistent)

### PR DESCRIPTION
**Pull request by ardera**
- remove buildWithAssemble shim since only BundleBuilder was using it
- move that functionality into the BundleBuilder itself
- source all the build parameters and build system defines from buildInfo instead of extra parameters
- make the bundle_shim_test.dart test the BundleBuilder instead (and rename it)
- migrate all the tests and usages (that's only the flutter build bundle command and CustomDevices)
- added one new test that makes sure all the right environment defines are passed to the build system

fixes #80732 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/bundle.dart [hunk 1]
```diff
@@ -84,34 +84,83 @@ class BundleBuilder {
   /// The default  `manifestPath` is `pubspec.yaml`
   Future<void> build({
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/bundle.dart:85]: You could arrange this a little bit differently to make the unit tests a bit easier to write for now (avoiding the top level `env` declaration.

Add a new optional parameter: `@visibleForTesting BuildSystem buildSystemOverride`

Then use it with something like `buildSystemOverride ?? globals.buildSystem`. You can then pass the build system directly to the bundle builder

Comment (ardera) [reply packages/flutter_tools/lib/src/bundle.dart:85]: done

**Comment by ardera** [comment]
I think the one failing test is a fluke, only moved the `env` variable into the `testUsingContext` closure and it worked fine before that

**Comment by jonahwilliams** [comment]
looks like a timeout rather than a flake

**Comment by jonahwilliams** [comment]
hmm but its an integration test, I'll give it a kick

File: packages/flutter_tools/test/general.shard/bundle_builder_test.dart [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+import '../src/common.dart';
+import '../src/context.dart';
+import '../src/test_build_system.dart';
+import '../src/testbed.dart';
+
+// Tests for BundleBuilder.
+void main() {
+  testUsingContext('Copies assets to expected directory after building', () async {
+    await BundleBuilder().build(
```
[the remark this answers is not in the document]
Comment (ardera) [reply packages/flutter_tools/test/general.shard/bundle_builder_test.dart:35]: yep forgot about that for this test, however I'm not sure I can save any overrides, `BundleBuilder.build` accesses some globals directly

Generally, what stuff should be passed via the app context and what via arguments? Are you trying to use arguments as much as possible, or environment or does it not matter? I mean theoretically I could also make `BundleBuilder` accept the `fileSystem`, `cache`, `processManager`, `buildContext` and so on via the constructor similiar to how all the Device-related (`CustomDevice` etc) classes do it

[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/bundle_builder_test.dart:35]: global context injection is the legacy method of providing dependencies. Generally we've been slowly, slowly migrating code to just use regular old constructor injection of deps, or to provide visibleForTesting arguments where globals are still used heavily. If you're refactoring or rewriting code, prefer restructuring it to not depend on context injection

[the remark this answers is not in the document]
Comment (ardera) [reply packages/flutter_tools/test/general.shard/bundle_builder_test.dart:35]: Ah okay. Yeah I might as well use constructor arguments then

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by ardera** [comment]
nice!

> looks like a timeout rather than a flake

oh, thought thats the same thing

**Pull request merged**